fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/join/JoinExchange.java

```diff
@@ -0,0 +1,136 @@
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package org.apache.doris.nereids.rules.exploration.join;
+
+import org.apache.doris.nereids.annotation.Developing;
+import org.apache.doris.nereids.rules.Rule;
+import org.apache.doris.nereids.rules.RuleType;
+import org.apache.doris.nereids.rules.exploration.OneExplorationRuleFactory;
+import org.apache.doris.nereids.trees.expressions.Expression;
+import org.apache.doris.nereids.trees.expressions.Slot;
+import org.apache.doris.nereids.trees.plans.GroupPlan;
+import org.apache.doris.nereids.trees.plans.JoinType;
+import org.apache.doris.nereids.trees.plans.Plan;
+import org.apache.doris.nereids.trees.plans.logical.LogicalJoin;
+import org.apache.doris.nereids.util.JoinUtils;
+
+import com.google.common.collect.Lists;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * rule factory for exchange without inside-project.
+ */
+@Developing
+public class JoinExchange extends OneExplorationRuleFactory {
+    public static final JoinExchange INSTANCE = new JoinExchange();
+
+    /*
+     *        topJoin                      newTopJoin
+     *        /      \                      /      \
+     *   leftJoin  rightJoin   -->   newLeftJoin newRightJoin
+     *    /    \    /    \            /    \        /    \
+     *   A      B  C      D          A      C      B      D
+     */
+    @Override
+    public Rule build() {
+        return innerLogicalJoin(innerLogicalJoin(), innerLogicalJoin())
+                .when(JoinExchange::checkReorder)
+                .then(topJoin -> {
+                    LogicalJoin<GroupPlan, GroupPlan> leftJoin = topJoin.left();
+                    LogicalJoin<GroupPlan, GroupPlan> rightJoin = topJoin.right();
+                    GroupPlan a = leftJoin.left();
+                    GroupPlan b = leftJoin.right();
+                    GroupPlan c = rightJoin.left();
+                    GroupPlan d = rightJoin.right();
+
+                    Set<Slot> acOutputSet = JoinUtils.getJoinOutput(a, c);
+                    Set<Slot> bdOutputSet = JoinUtils.getJoinOutput(b, d);
+
+                    List<Expression> newLeftJoinHashJoinConjuncts = Lists.newArrayList();
+                    List<Expression> newRightJoinHashJoinConjuncts = Lists.newArrayList();
+                    List<Expression> newTopJoinHashJoinConjuncts = new ArrayList<>(leftJoin.getHashJoinConjuncts());
+                    newTopJoinHashJoinConjuncts.addAll(rightJoin.getHashJoinConjuncts());
+                    splitTopConditon(topJoin.getHashJoinConjuncts(), acOutputSet, bdOutputSet,
+                            newLeftJoinHashJoinConjuncts, newRightJoinHashJoinConjuncts, newTopJoinHashJoinConjuncts);
```

Comment:
should we avoid generate cross join?

Reply:
I forgot to check empty list to avoid cross join.